pkg/codegen/hcl2/model/type_object.go

```diff
@@ -69,15 +69,25 @@ func (t *ObjectType) Traverse(traverser hcl.Traverser) (Traversable, hcl.Diagnos
 		return t.propertyUnion, nil
 	}
 
-	keyString, err := convert.Convert(key, cty.String)
-	contract.Assert(err == nil)
+	if keyType == NumberType {
 
-	propertyName := keyString.AsString()
-	propertyType, hasProperty := t.Properties[propertyName]
-	if !hasProperty {
-		return DynamicType, hcl.Diagnostics{unknownObjectProperty(propertyName, traverser.SourceRange())}
+		// The key is of NumberType, which means that it refers to the legacy HCL indexation syntax (".0", ".1", etc)
+		return t, nil
+	} else {
+
+		// The key is of PropertyType, which means that it is most likely a property (".skuName", "tenantId", etc)
+		keyString, err := convert.Convert(key, cty.String)
+		contract.Assert(err == nil)
+
+		propertyName := keyString.AsString()
+		propertyType, hasProperty := t.Properties[propertyName]
+		if !hasProperty {
+
+			// The specific key can't be found in our property list.
+			return DynamicType, hcl.Diagnostics{unknownObjectProperty(propertyName, traverser.SourceRange())}
+		}
+		return propertyType, nil
```

Comment:
Nit: it's idiomatic in Go to omit the `else` statement if the `else` follows an `if` that always returns.

```suggestion
	// The key is of PropertyType, which means that it is most likely a property (".skuName", "tenantId", etc)
	keyString, err := convert.Convert(key, cty.String)
	contract.Assert(err == nil)

	propertyName := keyString.AsString()
	propertyType, hasProperty := t.Properties[propertyName]
	if !hasProperty {
		// The specific key can't be found in our property list.
		return DynamicType, hcl.Diagnostics{unknownObjectProperty(propertyName, traverser.SourceRange())}
	}
	return propertyType, nil
```